### PR TITLE
Improve same_types and remove checking in backward

### PR DESCRIPTION
**Pull request by okuta**
This PR deletes type check at backward because it is redundant.

### CODE REVIEW

File: chainer/functions/connection/dilated_convolution_2d.py [hunk 1]
```diff
@@ -88,16 +88,6 @@ def forward_gpu(self, inputs):
...
-        if not type_check.same_types(*inputs):
-            if b is not None:
-                raise ValueError('numpy and cupy must not be used together\n'
-                                 'type(W): {0}, type(x): {1}, type(b): {2}'
-                                 .format(type(W), type(x), type(b)))
-            else:
-                raise ValueError('numpy and cupy must not be used together\n'
-                                 'type(W): {0}, type(x): {1}'
-                                 .format(type(W), type(x)))
-
```
Comment (rezoo) [inline chainer/functions/connection/dilated_convolution_2d.py]: Why you deleted the above lines?

File: chainer/utils/type_check.py [hunk 2]
```diff
@@ -515,15 +515,10 @@ def expect(*bool_exprs):
...
-    are_numpy_arrays = map(lambda x: issubclass(type(x), numpy.ndarray),
-                           arrays)
-    all_numpy_arrays = all(are_numpy_arrays)
-    if cuda.available:
-        are_cupy_arrays = map(lambda x: issubclass(type(x), cuda.cupy.ndarray),
-                              arrays)
-        return all_numpy_arrays or all(are_cupy_arrays)
-    else:
-        return all_numpy_arrays
+    if all(map(lambda x: isinstance(x, numpy.ndarray), arrays)):
```
Comment (rezoo) [inline chainer/utils/type_check.py]: How about using generator comprehension like that?
```python
if all(isinstance(x, numpy.ndarray) for x in arrays):
```

File: chainer/utils/type_check.py [hunk 3]
```diff
@@ -515,15 +515,10 @@ def expect(*bool_exprs):
...
+        return True
+    return cuda.available and all([isinstance(x, cuda.ndarray)
+                                   for x in arrays])
```
Comment (rezoo) [inline chainer/utils/type_check.py]: Ditto.
```python
return cuda.available and all(isinstance(x, cuda.ndarray) for x in arrays)
```

File: chainer/utils/type_check.py (hunk 2 above)
Comment (okuta) [inline chainer/utils/type_check.py]: Current code is faster than your suggested code.

Comment (rezoo) [inline chainer/utils/type_check.py]: Is it obvious? I'm just curious to see the difference.

**Comment by okuta** [comment]
I fixed!

**Comment by rezoo** [comment]
LGTM.

**Pull request merged**